File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    if (!s.length) {
        return 0;
    }
    let max = 0;
    let set = new Set();
    let right;
    let left = 0;

    for (right = 0; right < s.length; right++) {
        if (!set.has(s[right])) {
            set.add(s[right]);
        } else {
            while (set.has(s[right])) {
                set.delete(s[left]);
                left++;
            }
            set.add(s[right]);
        }
        
        max = Math.max(max, right - left + 1);
    }
    return max;
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        m = 0

        setter = set()

        left = 0
        for right in range(len(s)):
            while s[right] in setter:
                setter.remove(s[left])
                left += 1

            setter.add(s[right])
            
            m = max(m, right - left + 1)

        return m

## Step 1
## Determine what algos to use
## sliding window possibly hash table, using a set's size
## Variables to use (max, left pointer, right pointer)
## 